oraclepaas/resource_mysql_service_instance.go

```diff
@@ -0,0 +1,662 @@
+	attrs := mysqlConfiguration[0].(map[string]interface{})
+
+	// get the first entry.
+	MysqlInput := &mysql.MySQLParameters{
+		DBName:            attrs["db_name"].(string),
+		DBStorage:         strconv.Itoa(attrs["db_storage"].(int)),
+		MysqlUserName:     attrs["mysql_username"].(string),
+		MysqlUserPassword: attrs["mysql_password"].(string),
+	}
+
+	log.Printf("[DEBUG] Enterprise Monitor : %v", attrs["enterprise_monitor"])
+
+	if val, ok := attrs["enterprise_monitor"]; ok {
+		if val.(bool) == true {
+			MysqlInput.EnterpriseMonitor = "Yes"
+		} else {
+			MysqlInput.EnterpriseMonitor = "No"
+		}
+	}
+
+	err := expandEM(attrs, MysqlInput)
+	if err != nil {
+		return result, err
+	}
+
+	if val, ok := attrs["mysql_charset"]; ok && val != "" {
+		MysqlInput.MysqlCharset = val.(string)
+	}
+
+	if val, ok := attrs["mysql_collation"]; ok && val != "" {
+		MysqlInput.MysqlCollation = val.(string)
+	}
+
+	if val, ok := attrs["mysql_port"]; ok && val != "" {
+		MysqlInput.MysqlPort = strconv.Itoa(val.(int))
+	}
+
+	if val, ok := attrs["shape"]; ok && val != "" {
+		MysqlInput.Shape = val.(string)
+	}
+
+	if val, ok := attrs["snapshot_name"]; ok && val != "" {
+		MysqlInput.SnapshotName = val.(string)
+	}
+
+	if val, ok := attrs["source_service_name"]; ok && val != "" {
+		MysqlInput.SourceServiceName = val.(string)
+	}
+
+	if val, ok := attrs["subnet"]; ok && val != "" {
+		MysqlInput.Subnet = val.(string)
+	}
+
+	result.Mysql = *MysqlInput
+	return result, nil
+}
+
+func resourceOraclePAASMySQLServiceInstanceRead(d *schema.ResourceData, meta interface{}) error {
+
+	log.Printf("[DEBUG] Resource state: %#v", d.State())
+	mysqlClient, err := getMySQLClient(meta)
+	if err != nil {
+		return err
+	}
+	client := mysqlClient.ServiceInstanceClient()
+
+	input := mysql.GetServiceInstanceInput{
+		Name: d.Id(),
+	}
+
+	result, err := client.GetServiceInstance(&input)
+
+	if err != nil {
+		if opcClient.WasNotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading database service instance %s: %+v", d.Id(), err)
+	}
+
+	// if there is not result, there was an earlier issue. We set the ID of the mysql instance to blank.
+	if result == nil {
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] Read state of mysql service instance %s: %#v", d.Id(), result)
+
+	d.Set("service_name", result.ServiceName)
+	d.Set("service_description", result.ServiceDescription)
+	d.Set("backup_destination", result.BackupDestination)
+	d.Set("metering_frequency", result.MeteringFrequency)
+	d.Set("service_id", result.ServiceId)
+	d.Set("service_type", result.ServiceType)
+	d.Set("release_version", result.ReleaseVersion)
+	d.Set("service_version", result.ServiceVersion)
+	d.Set("base_release_version", result.BaseReleaseVersion)
+	d.Set("creator", result.Creator)
+	d.Set("creation_date", result.CreationDate)
+
+	if err := updateMySQLAttributesFromAttachments(d, result.Components.Mysql); err != nil {
```

Comment:
As we do expand for reading the terraform configuration, we like to call flatten for setting the terraform state so let's refactor this to `flattenMySQLAttributesFromAttachments`.

Reply:
Done